File: pcbstatorgen/stackup/power.py
"""
pcbstatorgen.stackup.power
===========================
Estimates power consumption, thermal rise, and capacitor bank sizing for the
stator drive circuit and produces a :class:`~pcbstatorgen.config.PowerBudget`.

Model overview
--------------
The power model proceeds in three steps:

1. **DC resistance** — computes the total trace length for one phase using
   the :class:`~pcbstatorgen.geometry.wave_winding.WaveWindingGenerator` at
   the given layer count, then applies
   :func:`~pcbstatorgen.units.cu_resistance_per_length`.

2. **Temperature rise** — uses the IPC-2152 empirical formula for PCB copper
   conductors in free air.  The aggregate board temperature rise (from all
   phases dissipating simultaneously) is estimated using a PCB thermal
   resistance of 15 °C/W — appropriate for a 20 mm × 200 mm board with
   natural convection.

3. **Burst / capacitor sizing** — the burst current for rapid automation
   moves is sourced from a local capacitor bank to avoid drooping the supply
   rail.  The minimum bank is sized to supply the burst current for 100 ms
   with ≤ 10 % voltage droop.

Efficiency note
---------------
The mechanical efficiency of a short-stroke coreless PCB mover at rated
continuous current is typically 1–5 %.  Most electrical energy is dissipated
as heat even at full speed because the mover spends most of its time in
position-hold (zero velocity → zero mechanical output).
"""

from __future__ import annotations

import math

from pcbstatorgen.config import LinearMotorConfig, PowerBudget, StackupResult
from pcbstatorgen.units import cu_resistance_per_length, oz_to_m

__all__ = ["PowerEstimator"]

#: PCB thermal resistance [°C/W] — 20 mm × 200 mm board, natural convection.
_R_THERMAL_C_PER_W: float = 15.0

#: Burst duration used for capacitor sizing [s].
_T_BURST_S: float = 0.1  # 100 ms

#: Maximum acceptable voltage droop during burst as fraction of supply voltage.
_DROOP_FRACTION: float = 0.10  # 10 %

#: Approximate rated velocity for efficiency calculation [m/s].
_V_RATED_M_S: float = 0.10


class PowerEstimator:
    """Estimate drive-circuit power budget for a mover stator design.

    The estimator can work with or without a :class:`~pcbstatorgen.config.StackupResult`:

    * **With** ``StackupResult`` — uses actual per-layer trace widths and
      copper thicknesses from the layer optimiser.
    * **Without** — falls back to conservative defaults: trace width =
      ``2 × min_trace_m``, inner copper = 2 oz.

    Parameters
    ----------
    layers_per_phase:
        How many copper layers carry each phase.  When ``None`` (default),
        the estimator infers this from the ``StackupResult`` if provided,
        or uses 2 as a fallback.

    Examples
    --------
    Quick estimate without a stackup::

        from pcbstatorgen.stackup.power import PowerEstimator

        est = PowerEstimator()
        budget = est.estimate(config)
        print(budget.summary())

    With LayerOptimizer output::

        budget = est.estimate(config, stackup_result=stackup)
    """

    def __init__(self, layers_per_phase: int | None = None) -> None:
        if layers_per_phase is not None and layers_per_phase < 1:
            raise ValueError(
                f"layers_per_phase must be ≥ 1, got {layers_per_phase}"
            )
        self.layers_per_phase = layers_per_phase

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def estimate(
        self,
        config: LinearMotorConfig,
        stackup_result: StackupResult | None = None,
    ) -> PowerBudget:
        """Compute a :class:`~pcbstatorgen.config.PowerBudget`.

        Parameters
        ----------
        config:
            Linear motor configuration.
        stackup_result:
            Optional stackup from the LayerOptimizer.  If provided, actual
            per-layer trace widths and copper weights are used.

        Returns
        -------
        PowerBudget
        """
        trace_width_m, cu_thickness_m, lpp = self._trace_params(
            config, stackup_result
        )

        # Coil total length per layer per phase (active + end-turn)
        single_layer_length_m = self._single_layer_trace_length_m(config)
        total_length_m = single_layer_length_m * lpp

        # DC resistance per phase
        R_per_m = cu_resistance_per_length(trace_width_m, cu_thickness_m)
        R_phase = R_per_m * total_length_m

        # Continuous power — all phases on simultaneously
        I_cont = config.max_current_a
        P_cont = config.phases * I_cont ** 2 * R_phase

        # Burst current — scales linearly with force ratio
        if config.target_force_n > 0:
            I_burst = I_cont * (config.peak_force_n / config.target_force_n)
        else:
            I_burst = I_cont
        P_burst = config.phases * I_burst ** 2 * R_phase

        # Temperature rise (aggregate)
        delta_T = P_cont * _R_THERMAL_C_PER_W

        # Capacitor sizing
        delta_V = config.supply_voltage_v * _DROOP_FRACTION
        C_uf = max(
            0.0,
            (I_burst * config.phases * _T_BURST_S / delta_V) * 1e6,
        )

        # Mechanical efficiency at rated velocity
        P_mech = config.target_force_n * _V_RATED_M_S
        P_elec = config.supply_voltage_v * I_cont
        efficiency = min(100.0, (P_mech / P_elec * 100.0) if P_elec > 0 else 0.0)

        return PowerBudget(
            phase_resistance_ohm=R_phase,
            continuous_power_w=P_cont,
            burst_power_w=P_burst,
            temperature_rise_c=delta_T,
            capacitor_required_uf=C_uf,
            efficiency_pct=efficiency,
        )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _trace_params(
        self,
        config: LinearMotorConfig,
        stackup_result: StackupResult | None,
    ) -> tuple[float, float, int]:
        """Return ``(trace_width_m, cu_thickness_m, layers_per_phase)``."""
        if self.layers_per_phase is not None:
            lpp = self.layers_per_phase
        elif stackup_result is not None:
            lpp = max(1, stackup_result.layer_count // config.phases)
        else:
            lpp = 2  # conservative fallback

        if stackup_result is not None:
            # Use inner-layer values (index 1) for the dominant resistance contribution
            idx = min(1, len(stackup_result.trace_widths_m) - 1)
            tw = stackup_result.trace_widths_m[idx]
            ct = stackup_result.cu_thickness_m[idx]
        else:
            # Conservative defaults: 2× min trace, 2 oz copper
            tw = config.min_trace_m * 2.0
            ct = oz_to_m(2.0)

        return tw, ct, lpp

    @staticmethod
    def _single_layer_trace_length_m(config: LinearMotorConfig) -> float:
        """Approximate total trace length per phase per layer [m].

        Uses the WaveWindingGenerator to compute the exact value for the
        SERPENTINE topology.  Other topologies are within ~15 % of this.
        """
        from pcbstatorgen.geometry.wave_winding import WaveWindingGenerator
        gen = WaveWindingGenerator()
        coils = gen.generate(config, layer_idx=0)
        if not coils:
            return 0.0
        # Average across all phases (they differ by at most 2 conductors)
        return sum(c.total_length_m for c in coils) / len(coils)

File: crates/pcbstatorgen-rs/src/stackup/power.rs
//! Power / thermal estimator. Ports `pcbstatorgen/stackup/power.py`.

use crate::config::{LinearMotorConfig, PowerBudget, StackupResult};
// TODO(product-owner): re-import WaveWindingGenerator once geometry Phase C
//       restructuring is complete. Temporarily using a simple approximation.
// use crate::geometry::wave_winding::WaveWindingGenerator;
use crate::units::{cu_resistance_per_length, oz_to_m, RHO_CU};

/// PCB thermal resistance [°C/W] — 20 mm × 200 mm board, natural convection.
const R_THERMAL_C_PER_W: f64 = 15.0;
/// Burst duration for capacitor sizing [s].
const T_BURST_S: f64 = 0.1;
/// Max acceptable voltage droop during burst (fraction of supply).
const DROOP_FRACTION: f64 = 0.10;
/// Approximate rated velocity for efficiency [m/s].
const V_RATED_M_S: f64 = 0.10;

/// Estimate drive-circuit power budget.
#[derive(Debug, Clone)]
pub struct PowerEstimator {
    /// How many copper layers carry each phase.
    pub layers_per_phase: Option<u32>,
}

impl Default for PowerEstimator {
    fn default() -> Self {
        Self { layers_per_phase: None }
    }
}

impl PowerEstimator {
    /// Compute a PowerBudget.
    pub fn estimate(
        &self,
        config: &LinearMotorConfig,
        stackup: Option<&StackupResult>,
    ) -> PowerBudget {
        let (trace_width_m, cu_thickness_m, lpp) = self.trace_params(config, stackup);

        let single_layer_length_m = Self::single_layer_trace_length_m(config);
        let total_length_m = single_layer_length_m * lpp as f64;

        let r_per_m = cu_resistance_per_length(trace_width_m, cu_thickness_m, RHO_CU);
        let r_phase = r_per_m * total_length_m;

        let i_cont = config.max_current_a;
        let p_cont = config.phases as f64 * i_cont * i_cont * r_phase;

        let i_burst = if config.target_force_n > 0.0 {
            i_cont * (config.peak_force_n / config.target_force_n)
        } else {
            i_cont
        };
        let p_burst = config.phases as f64 * i_burst * i_burst * r_phase;

        let delta_t = p_cont * R_THERMAL_C_PER_W;

        let delta_v = config.supply_voltage_v * DROOP_FRACTION;
        let c_uf = if delta_v > 0.0 {
            ((i_burst * config.phases as f64 * T_BURST_S / delta_v) * 1e6).max(0.0)
        } else {
            0.0
        };

        let p_mech = config.target_force_n * V_RATED_M_S;
        let p_elec = config.supply_voltage_v * i_cont;
        let efficiency = if p_elec > 0.0 {
            (p_mech / p_elec * 100.0).min(100.0)
        } else {
            0.0
        };

        PowerBudget {
            phase_resistance_ohm: r_phase,
            continuous_power_w: p_cont,
            burst_power_w: p_burst,
            temperature_rise_c: delta_t,
            capacitor_required_uf: c_uf,
            efficiency_pct: efficiency,
        }
    }

    fn trace_params(
        &self,
        config: &LinearMotorConfig,
        stackup: Option<&StackupResult>,
    ) -> (f64, f64, u32) {
        let lpp = if let Some(l) = self.layers_per_phase {
            l
        } else if let Some(s) = stackup {
            ((s.layer_count / config.phases) as u32).max(1)
        } else {
            2
        };

        if let Some(s) = stackup {
            let idx = (1).min(s.trace_widths_m.len() - 1);
            let tw = s.trace_widths_m[idx];
            let ct = s.cu_thickness_m[idx];
            (tw, ct, lpp)
        } else {
            (config.min_trace_m * 2.0, oz_to_m(2.0), lpp)
        }
    }

    /// Approximate total trace length per phase per layer [m].
    fn single_layer_trace_length_m(config: &LinearMotorConfig) -> f64 {
        // TODO(product-owner): restore real WaveWindingGenerator call once
        //       geometry Phase C restructuring is complete.
        // Approximate: 2 * board_width per conductor, ~magnet_count*2 conductors.
        let n_conductors = (config.magnet_count * 2) as f64;
        2.0 * config.board_width_m * n_conductors
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::StackupResult;
    use crate::units::{mm, oz_to_m};

    fn default_config() -> LinearMotorConfig {
        LinearMotorConfig {
            active_area_length_m: mm(195.0),
            magnet_dims_m: [mm(10.0), mm(10.0), mm(4.0)],
            magnet_count: 10,
            magnet_pitch_m: mm(12.0),
            phases: 3,
            target_force_n: 0.5,
            peak_force_n: 1.0,
            max_current_a: 1.0,
            supply_voltage_v: 5.0,
            min_trace_m: mm(0.127),
            min_space_m: mm(0.127),
            board_width_m: mm(20.0),
            air_gap_m: mm(0.5),
            ..LinearMotorConfig::default()
        }
    }

    fn four_layer_stackup() -> StackupResult {
        StackupResult {
            layer_count: 4,
            trace_widths_m: vec![mm(0.15), mm(0.25), mm(0.25), mm(0.15)],
            cu_thickness_m: vec![oz_to_m(1.0), oz_to_m(2.0), oz_to_m(2.0), oz_to_m(1.0)],
            via_drill_m: mm(0.2),
            via_annular_ring_m: mm(0.1),
            via_grid_rows: 2,
            via_grid_cols: 3,
            estimated_force_n: 0.42,
            estimated_dc_resistance_ohm: 3.1,
            notes: vec![],
        }
    }

    #[test]
    fn test_estimate_without_stackup() {
        let cfg = default_config();
        let est = PowerEstimator::default();
        let pb = est.estimate(&cfg, None);
        assert!(pb.phase_resistance_ohm > 0.0);
        assert!(pb.continuous_power_w > 0.0);
        assert!(pb.burst_power_w >= pb.continuous_power_w);
        assert!(pb.temperature_rise_c > 0.0);
        assert!(pb.efficiency_pct >= 0.0 && pb.efficiency_pct <= 100.0);
    }

    #[test]
    fn test_estimate_with_stackup() {
        let cfg = default_config();
        let s = four_layer_stackup();
        let est = PowerEstimator::default();
        let pb = est.estimate(&cfg, Some(&s));
        assert!(pb.phase_resistance_ohm > 0.0);
    }

    #[test]
    fn test_burst_higher_than_continuous() {
        let cfg = default_config();
        let est = PowerEstimator::default();
        let pb = est.estimate(&cfg, None);
        // peak_force = 1.0, target = 0.5 → I_burst = 2×I_cont → P_burst = 4×P_cont
        assert!(pb.burst_power_w > pb.continuous_power_w);
    }

    #[test]
    fn test_capacitor_positive() {
        let cfg = default_config();
        let est = PowerEstimator::default();
        let pb = est.estimate(&cfg, None);
        assert!(pb.capacitor_required_uf > 0.0);
    }

    #[test]
    fn test_custom_layers_per_phase() {
        let cfg = default_config();
        let est = PowerEstimator { layers_per_phase: Some(1) };
        let pb_default = PowerEstimator::default().estimate(&cfg, None);
        let pb_custom = est.estimate(&cfg, None);
        // With 1 layer per phase vs 2, resistance should be 2× and power 2×
        assert!(pb_custom.phase_resistance_ohm < pb_default.phase_resistance_ohm);
    }
}
